nginx service: add nested locations support

sshisk opened the pull request:
###### Motivation for this change

Copy locations from vhost-options.nix so can be nested.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Infinisil commented on lib/types.nix:
```diff
@@ -456,6 +456,12 @@ rec {
+      getSubOptions = prefix: {};
+      getSubModules = null;
+    };
+
```
You can just override the description of the type (`type // { description = "custom descr"; }`) to get rid of the infinite recursion, no function in lib needed.

sshisk replied:
This doesn't work. Recursion occurs in manual generation when it tries to generate options description for `locations._name_.locations._name_` etc.

roberth replied:
There's potentially a more thorough solution now, in 
 - https://github.com/NixOS/nixpkgs/pull/208227

Look for `recursive =` in `types.nix`.

cc @ncfavier

infinisil reviewed (changes requested):
What are nested locations used for?

sshisk commented:
@Infinisil I use them to restrict access to sublocations with password. They are useful for other settings too.

sshisk commented:
Any update?

Infinisil commented on nixos/modules/services/web-servers/nginx/location-options.nix:
```diff
@@ -7,6 +7,7 @@
 
 with lib;
 
+let self =
```
Why this?

sshisk replied:
You are right, I forgot to actually use `self`. Fixed that.

abbradar commented:
I have used this for sub-locations like those:

```
          locations = {
            "/private/" = {
              extraConfig = ''
                auth_basic "Secret area";
                auth_basic_user_file /root/private-htpasswd;
              '';
              locations."/private/files/".extraConfig = "autoindex on;";
            };
```

sshisk committed:
hideSubTree: add to break recursion

This function is helpful for infinitely recursive types to stop traversing early.
Without this infinitely recursive options crash Nix during documentation build.

sshisk committed:
nginx service: add nested locations support

sshisk commented:
I rebased tree atop latest master.

abbradar commented:
@Infinisil kind ping -- I'd like to merge because I use this patch, but not sure you are okay with this. Also I'm not sure `hideSubTree` is the good way here.

Infinisil commented on nixos/modules/services/web-servers/nginx/location-options.nix:
```diff
@@ -111,6 +112,21 @@ with lib;
     };
 
+    locations = mkOption {
+      type = types.attrsOf (types.hideSubTree (types.submodule self));
```
The `hideSubTree` hack is not necessary anymore thanks to @roberth with https://github.com/NixOS/nixpkgs/pull/143207/commits/64dfd983df61d9c427ac01d0d25b8d2f26b4c3d5. Now you can do

```suggestion
      type = types.attrsOf (types.submodule self);
      visible = "shallow";
```

Infinisil commented on nixos/modules/services/web-servers/nginx/location-options.nix:
```diff
@@ -7,6 +7,7 @@
 
 with lib;
 
+let self =
```
Suggestion: Instead of doing the recursion here, do it in `./vhost-options.nix` by doing

```diff
diff --git a/nixos/modules/services/web-servers/nginx/vhost-options.nix b/nixos/modules/services/web-servers/nginx/vhost-options.nix
index c4e8285dc48..d687cb8ab2d 100644
--- a/nixos/modules/services/web-servers/nginx/vhost-options.nix
+++ b/nixos/modules/services/web-servers/nginx/vhost-options.nix
@@ -271,9 +271,12 @@ with lib;
     };
 
     locations = mkOption {
-      type = types.attrsOf (types.submodule (import ./location-options.nix {
-        inherit lib;
-      }));
+      type =
+        let
+          locationsType = types.attrsOf (types.submodule (import ./location-options.nix {
+            inherit lib locationsType;
+          }));
+        in locationsType;
       default = {};
       example = literalExpression ''
         {
```

Then you can take the `locationsType` argument in this module (though it's passed a bit unconventional, but that's another issue), and use that directly as the type for the `locations` option.

infinisil reviewed (commented):
I'd like to merge this once these points are addressed :)

kirillrdy commented:
please resolve conflicts

infinisil commented:
It looks like the PR author isn't active anymore. I encourage others to pick this up.
